File: 649-dota2-senate/solution.py
from collections import deque

class Solution:
    # Recommended approach uses queue. Use 'deque' for O(1) dequeue using popleft()
    def predictPartyVictory(self, senate: str) -> str:
        n = len(senate)
        r = deque()
        d = deque()

        # store all the senators into their respective party's queue, but store their indexes, the count is length of queue
        for idx, i in enumerate(senate):
            if i == 'D':
                d.append(idx)
            else:
                r.append(idx)

        # loop until one of the queues is empty
        while r and d:
            r_idx = r.popleft()
            d_idx = d.popleft()
            # check which senator's index is smaller (aka which comes first in the line, and takes priority)
            if r_idx < d_idx:
                # if rad goes first, add it to the back of the queue, and the dire that lost the condition stays out...
                r.append(r_idx + n)
            else:
                d.append(d_idx + n)
        
        return "Radiant" if r else "Dire"

File: 649-dota2-senate/solution.go
package main

type Queue []int

func (q *Queue) Append(i int) {
	*q = append(*q, i)
}

func (q *Queue) Popleft() int {
	if len(*q) == 0 {
		return 99999999
	}
	i := (*q)[0]
	*q = (*q)[1:]
	return i
}

func predictPartyVictory(senate string) string {
	n := len(senate)
	var r Queue
	var d Queue
	for idx, i := range senate {
		if i == 'D' {
			d.Append(idx)
		} else {
			r.Append(idx)
		}
	}

	for len(r) > 0 && len(d) > 0 {
		r_idx := r.Popleft()
		d_idx := d.Popleft()
		if r_idx < d_idx {
			r.Append(r_idx + n)
		} else {
			d.Append(d_idx + n)
		}
	}

	if len(r) > 0 {
		return "Radiant"
	} else {
		return "Dire"
	}
}
